[JSC] Skip ProxyObject's ICs trap result validation in the common case

shvaikalesh opened the pull request:
#### 18d9fef12db3ad57b2a20f7a8c8120059be66c70
<pre>
[JSC] Skip ProxyObject&apos;s ICs trap result validation in the common case
<a href="https://bugs.webkit.org/show_bug.cgi?id=255661">https://bugs.webkit.org/show_bug.cgi?id=255661</a>
&lt;rdar://problem/108269769&gt;

Reviewed by Yusuke Suzuki.

Invariants of the essential internal methods [1] are enforced for all objects, both ordinary and exotic,
to avoid them having unpredictable behavior that negatively impacts developer experience and creates
security issues.

Enforcing them for Proxy objects is especially cumbersome, given traps are userland code, and quite
expensive performance-wise: invariants invalidation requires checking property attributes,
which currently requires JS =&gt; C++ call.

After carefully analyzing all the invariants of the all internal methods, it was concluded that if
[[ProxyTarget]] is an ordinary extensible object without non-configurable properties, validation of
the invariants can be safely (non-observably) skipped.

trap name                   IC    attributes checked
========================    ==    ==================================
getOwnPropertyDescriptor    no    non-configurable || non-extensible
defineProperty              no    non-configurable || non-extensible
has                        yes    non-configurable || non-extensible
get                        yes    non-configurable &amp;&amp; (non-writable || accessor)
set                        yes    non-configurable &amp;&amp; (non-writable || accessor)
deleteProperty           maybe    non-configurable || non-extensible
ownKeys                     no    non-configurable || non-extensible

This change:

  1. Introduces OverridesIsExtensible out-of-line type info flag to tighten Structure&apos;s didPreventExtensions()
     flag, making it more pessimistic: it being `true` doesn&apos;t necessarily mean the structure is non-extensible,
     but the reverse of it being `false` can be safely relied on.

     It&apos;s not a completely new approach: <a href="https://github.com/WebKit/WebKit/pull/11843">https://github.com/WebKit/WebKit/pull/11843</a> does the same thing.

     This change was proved not to introduce bugs since Structure&apos;s didPreventExtensions() flag is currently
     checked only for objects without isExtensible() override.

  2. Introduces hasNonConfigurableProperties and hasNonConfigurableReadOnlyOrGetterSetterProperties Structure
     flags that are also pessimistic.

     When inferring their initial values, we do extra care not to be overly pessimistic for objects except
     JSFinalObject by checking overridesGetOwnPropertySlot() so that userland code, which heavily uses
     Map / Set as [[ProxyTarget]], more performant, while making an exception for JSArray&apos;s non-configurable
     non-structure &quot;length&quot; property.

     While there is existing hasReadOnlyOrGetterSetterPropertiesExcludingProto flag, we can&apos;t combine it with
     hasNonConfigurableProperties because MobX installs a non-writable administration property even on arrays,
     which have non-configurable &quot;length&quot; already, so combining them would disable optimization for JSArray.

  3. Introduces HasStructureWithFlags bytecode op to be utilized in ProxyObject IC helpers to skip trap result
     validation for most of ProxyObject&apos;s targets. In the follow-ups, this approach will be expanded to C++ code.

Currently, due to lack of proper profiling of GetInternalField results, constant folding for HasStructureWithFlags
doesn&apos;t work in microbenchmarks.

                                   ToT                      patch

proxy-has-miss              159.9825+-2.4190     ^     43.1435+-0.5994        ^ definitely 3.7082x faster
proxy-set-miss-handler      158.7617+-2.3615     ^    139.2158+-1.9197        ^ definitely 1.1404x faster
proxy-set                   125.5967+-1.3678     ^     14.4533+-0.1739        ^ definitely 8.6898x faster
proxy-get                   115.8061+-2.2413     ^     14.2136+-0.1138        ^ definitely 8.1475x faster

&lt;geometric&gt;                 138.5985+-0.9526     ^     33.3237+-0.2255        ^ definitely 4.1592x faster

[1]: <a href="https://tc39.es/ecma262/#sec-invariants-of-the-essential-internal-methods">https://tc39.es/ecma262/#sec-invariants-of-the-essential-internal-methods</a>

* Source/JavaScriptCore/builtins/BuiltinNames.h:
* Source/JavaScriptCore/builtins/ProxyHelpers.js:
(linkTimeConstant.performProxyObjectHas):
(linkTimeConstant.performProxyObjectGet):
(linkTimeConstant.performProxyObjectSetSloppy):
(linkTimeConstant.performProxyObjectSetStrict):
* Source/JavaScriptCore/bytecode/BytecodeIntrinsicRegistry.h:
* Source/JavaScriptCore/bytecode/BytecodeList.rb:
* Source/JavaScriptCore/bytecode/BytecodeUseDef.cpp:
(JSC::computeUsesForBytecodeIndexImpl):
(JSC::computeDefsForBytecodeIndexImpl):
* Source/JavaScriptCore/bytecode/LinkTimeConstant.h:
* Source/JavaScriptCore/bytecompiler/BytecodeGenerator.cpp:
(JSC::BytecodeGenerator::emitHasStructureWithFlags):
* Source/JavaScriptCore/bytecompiler/BytecodeGenerator.h:
* Source/JavaScriptCore/bytecompiler/NodesCodegen.cpp:
(JSC::BytecodeIntrinsicNode::emit_intrinsic_mustValidateResultOfProxyGetAndSetTraps):
(JSC::BytecodeIntrinsicNode::emit_intrinsic_mustValidateResultOfProxyTrapsExceptGetAndSet):
* Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
(JSC::DFG::AbstractInterpreter&lt;AbstractStateType&gt;::executeEffects):
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::parseBlock):
* Source/JavaScriptCore/dfg/DFGClobberize.h:
(JSC::DFG::clobberize):
* Source/JavaScriptCore/dfg/DFGConstantFoldingPhase.cpp:
(JSC::DFG::ConstantFoldingPhase::foldConstants):
* Source/JavaScriptCore/dfg/DFGDoesGC.cpp:
(JSC::DFG::doesGC):
* Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
(JSC::DFG::FixupPhase::fixupNode):
* Source/JavaScriptCore/dfg/DFGHeapLocation.cpp:
(WTF::printInternal):
* Source/JavaScriptCore/dfg/DFGHeapLocation.h:
* Source/JavaScriptCore/dfg/DFGNode.h:
(JSC::DFG::Node::structureFlags):
* Source/JavaScriptCore/dfg/DFGNodeType.h:
* Source/JavaScriptCore/dfg/DFGPredictionPropagationPhase.cpp:
* Source/JavaScriptCore/dfg/DFGSafeToExecute.h:
(JSC::DFG::safeToExecute):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.h:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT32_64.cpp:
(JSC::DFG::SpeculativeJIT::compile):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
(JSC::DFG::SpeculativeJIT::compile):
* Source/JavaScriptCore/ftl/FTLCapabilities.cpp:
(JSC::FTL::canCompile):
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileNode):
(JSC::FTL::DFG::LowerDFGToB3::compileCompareStrictEq):
* Source/JavaScriptCore/jit/JIT.cpp:
(JSC::JIT::privateCompileMainPass):
* Source/JavaScriptCore/jit/JIT.h:
* Source/JavaScriptCore/jit/JITOpcodes.cpp:
(JSC::JIT::emit_op_has_structure_with_flags):
* Source/JavaScriptCore/llint/LowLevelInterpreter32_64.asm:
* Source/JavaScriptCore/llint/LowLevelInterpreter64.asm:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::init):
* Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::globalFuncHandleProxySetTrapResult): Deleted.
* Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.h:
* Source/JavaScriptCore/runtime/JSTypeInfo.h:
(JSC::TypeInfo::overridesIsExtensible const):
* Source/JavaScriptCore/runtime/PropertySlot.h:
* Source/JavaScriptCore/runtime/ProxyObject.cpp:
(JSC::ProxyObject::performPut):
(JSC::ProxyObject::validatePositiveSetTrapResult):
(JSC::ProxyObject::validateSetTrapResult): Deleted.
* Source/JavaScriptCore/runtime/ProxyObject.h:
* Source/JavaScriptCore/runtime/Structure.cpp:
(JSC::Structure::validateFlags):
(JSC::Structure::Structure):
(JSC::Structure::nonPropertyTransitionSlow):
* Source/JavaScriptCore/runtime/Structure.h:
(JSC::Structure::isNonExtensibleOrHasNonConfigurableProperties const):
(JSC::Structure::bitField const):
* Source/JavaScriptCore/runtime/StructureInlines.h:
(JSC::Structure::add):
(JSC::Structure::attributeChange):
* Source/JavaScriptCore/wasm/js/WebAssemblyGCObjectBase.h:
* Source/WebCore/bindings/js/JSDOMWindowProperties.h:

Canonical link: <a href="https://commits.webkit.org/263443@main">https://commits.webkit.org/263443@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a2d277d9c0013d21e64e8e66674899c188c88688

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/4639 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/4757 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/4915 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/6134 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/4786 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/4628 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/4898 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/4719 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/5028 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/4709 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/4797 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/4157 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/6143 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/2285 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/4149 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/9130 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/3844 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/4145 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/4223 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/5769 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/4401 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/4615 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/3752 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/4733 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/4140 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/1175 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/8186 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/4847 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/526 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/4500 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/1282 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
```diff
@@ -5969,6 +5969,21 @@ void SpeculativeJIT::compileIsTypedArrayView(Node* node)
+    test32(NonZero, Address(resultGPR, Structure::bitFieldOffset()), TrustedImm32(node->structureFlags()), resultGPR);
+
+    blessBoolean(resultGPR);
+    blessedBooleanResult(resultGPR, node);
```
Why not just use `unblessedBooleanResult`?

Constellation commented on Source/JavaScriptCore/bytecompiler/NodesCodegen.cpp:
```diff
@@ -2027,6 +2027,29 @@ CREATE_INTRINSIC_FOR_BRAND_CHECK(isUndefinedOrNull, IsUndefinedOrNull)
+
+    unsigned hasNonConfigurableProperties = 1 << Structure::s_hasNonConfigurablePropertiesShift;
+    unsigned didPreventExtensions = 1 << Structure::s_didPreventExtensionsShift;
+
```
Use `Structure::s_hasNonConfigurablePropertiesBits` etc.

Constellation commented on Source/JavaScriptCore/bytecompiler/NodesCodegen.cpp:
```diff
@@ -2027,6 +2027,29 @@ CREATE_INTRINSIC_FOR_BRAND_CHECK(isUndefinedOrNull, IsUndefinedOrNull)
+    RefPtr<RegisterID> src = generator.emitNode(node);
+    ASSERT(!node->m_next);
+
+    unsigned hasNonConfigurableReadOnlyOrGetterSetterProperties = 1 << Structure::s_hasNonConfigurableReadOnlyOrGetterSetterPropertiesShift;
```
Ditto

Constellation commented on Source/JavaScriptCore/dfg/DFGConstantFoldingPhase.cpp:
```diff
@@ -1094,6 +1094,28 @@ class ConstantFoldingPhase : public Phase {
+                if (JSValue value = abstractValue.value()) {
+                    m_graph.convertToConstant(node, jsBoolean(asObject(value)->structure()->bitField() & flags));
+                    break;
+                }
```
This is not correct. Object's constant's structure can be changed, so while the constant object is the same, it is possible that it gets different structure.

Constellation commented on Source/JavaScriptCore/dfg/DFGConstantFoldingPhase.cpp:
```diff
@@ -1094,6 +1094,28 @@ class ConstantFoldingPhase : public Phase {
+                    break;
+                }
+
+                if (abstractValue.m_structure.isFinite()) {
```
Let's first check that `abstractValue.m_type && !(abstractValue.m_type & ~SpecCell)`

Constellation commented on Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
```diff
@@ -4819,6 +4819,15 @@ bool AbstractInterpreter<AbstractStateType>::executeEffects(unsigned clobberLimi
 
+    case HasStructureWithFlags: {
+        AbstractValue& child = forNode(node->child1());
+        if (child.value() || child.m_structure.isFinite())
```
`child.value()` is wrong (I described why it is wrong in ConstantFoldingPhase's comment).
And we should check,
```
if (abstractValue.m_type && !(abstractValue.m_type & ~SpecCell) && abstractValue.m_structure.isFinite())
```

Constellation commented on Source/JavaScriptCore/dfg/DFGNode.h:
```diff
@@ -1486,6 +1486,12 @@ struct Node {
+    {
+        ASSERT(op() == HasStructureWithFlags);
+        return m_opInfo.as<uint32_t>();
+    }
```
You should add `bool hasStructureFlags() const` method too. And you should add flag dump support in DFGGraph.

Constellation commented on Source/JavaScriptCore/dfg/DFGClobberize.h:
```diff
@@ -1247,6 +1247,11 @@ void clobberize(Graph& graph, Node* node, const ReadFunctor& read, const WriteFu
 
+    case HasStructureWithFlags:
+        read(JSCell_structureID);
+        def(HeapLocation(HasStructureWithFlagsLoc, JSCell_structureID, node->child1()), LazyNode(node));
```
This define is not correct because even though JSCell_structureID is not written, the result of this can be changed (because these flags can be changed without structure transitions).

Constellation commented on Source/JavaScriptCore/dfg/DFGConstantFoldingPhase.cpp:
```diff
@@ -1117,6 +1117,24 @@ class ConstantFoldingPhase : public Phase {
+                ASSERT(flags);
+
+                if (abstractValue.m_type && !(abstractValue.m_type & ~SpecCell) && abstractValue.m_structure.isFinite()) {
+                    dataLogLn("CF");
```
dataLog remains.

shvaikalesh replied:
Oops sorry

Constellation commented on Source/JavaScriptCore/dfg/DFGConstantFoldingPhase.cpp:
```diff
@@ -1117,6 +1117,24 @@ class ConstantFoldingPhase : public Phase {
+                    abstractValue.m_structure.forEach([&] (RegisteredStructure structure) {
+                        anyStructureHasAnyFlag |= structure->hasAnyOfBitFieldFlags(flags);
+                    });
+                    m_graph.convertToConstant(node, jsBoolean(anyStructureHasAnyFlag));
```
Converting it to `jsBoolean(anyStructureHasAnyFlag)` looks wrong, right?
If one structure does not have this flag, but if the other candidate structures have, then it becomes true, and this is not correct.
You can fold to a constant only when all the structure agrees one result.

Constellation replied:
Also, you need to `change = true` when you changed the graph.

shvaikalesh replied:
Yeah my bad, changed, thanks!

Constellation commented on Source/JavaScriptCore/dfg/DFGConstantFoldingPhase.cpp:
```diff
@@ -1117,6 +1117,26 @@ class ConstantFoldingPhase : public Phase {
+                        m_graph.convertToConstant(node, jsBoolean(false));
+                        changed = true;
+                    }
+                }
```
Actually, this code is a bit wrong since Structure can change their flags without transitions. For example, `hasBeenFlattenedBefore` flag can be set without transitions. `didWatchReplacement` too. So this code is not correct since its flag is too generic. Some flags cannot be set without transition, but some are not. But right now, this code strongly assuming that these flags will not be modified.
I think you need to allow this folding only for specific Structure flags.

shvaikalesh replied:
Oh wow that's a great catch, introduced `Structure::bitFieldFlagsCantBeChangedWithoutTransition` to filter out those flags and even conditionally brought back removed clobberize based on it.

Constellation commented on Source/JavaScriptCore/dfg/DFGClobberize.h:
```diff
@@ -1247,6 +1247,20 @@ void clobberize(Graph& graph, Node* node, const ReadFunctor& read, const WriteFu
+            return;
+        }
+
+        clobberTop();
```
We are not writing anything. So, the definition should be readl(World).

shvaikalesh replied:
Changed 🙇

Constellation commented on Source/JavaScriptCore/dfg/DFGClobberize.h:
```diff
@@ -1247,6 +1247,20 @@ void clobberize(Graph& graph, Node* node, const ReadFunctor& read, const WriteFu
+            read(JSCell_structureID);
+            def(HeapLocation(HasStructureWithFlagsLoc, JSCell_structureID, node->child1()), LazyNode(node));
+            return;
+        }
```
This is not correct if the Structure is Dictionary.

shvaikalesh replied:
Fixed, thanks!

Constellation commented on Source/JavaScriptCore/dfg/DFGConstantFoldingPhase.cpp:
```diff
@@ -1117,6 +1117,26 @@ class ConstantFoldingPhase : public Phase {
+                unsigned flags = node->structureFlags();
+                ASSERT(flags);
+
+                if (Structure::bitFieldFlagsCantBeChangedWithoutTransition(flags) && abstractValue.m_type && !(abstractValue.m_type & ~SpecCell) && abstractValue.m_structure.isFinite()) {
```
This is not correct if Structure is Dictionary, correct?

shvaikalesh replied:
Yes, added an `isDictionary()` check to CF.

Constellation commented on Source/JavaScriptCore/runtime/Structure.h:
```diff
@@ -834,16 +844,31 @@ class Structure : public JSCell {
+            s_didPreventExtensionsBits
+            | s_isQuickPropertyAccessAllowedForEnumerationBits
+            | s_hasAnyKindOfGetterSetterPropertiesBits
+            | s_hasReadOnlyOrGetterSetterPropertiesExcludingProtoBits
+            | s_hasUnderscoreProtoPropertyExcludingOriginalProtoBits
+            | s_hasNonConfigurablePropertiesBits
+            | s_hasNonConfigurableReadOnlyOrGetterSetterPropertiesBits
```
They can be modified if the structure is Dictionary, correct?

shvaikalesh replied:
Yes, added an `isDictionary()` check to CF.

Constellation commented on Source/JavaScriptCore/dfg/DFGConstantFoldingPhase.cpp:
```diff
@@ -1118,6 +1118,25 @@ class ConstantFoldingPhase : public Phase {
+                if (Structure::bitFieldFlagsCantBeChangedWithoutTransition(flags) && child.m_type && !(child.m_type & ~SpecCell) && child.m_structure.isFinite()) {
+                    bool canFoldToFalse = true;
+                    child.m_structure.forEach([&] (RegisteredStructure structure) {
+                        canFoldToFalse &= !structure->isDictionary() && !structure->hasAnyOfBitFieldFlags(flags);
```
We should try folding it to true too. If all structure says `true` for the result, then we can fold it to true.
If some says false and some says true, then we should stop folding.

shvaikalesh replied:
Done, thank you!

Constellation reviewed (approved):
r=me with comments

Commit:
[JSC] Skip ProxyObject's ICs trap result validation in the common case
https://bugs.webkit.org/show_bug.cgi?id=255661
<rdar://problem/108269769>

Reviewed by Yusuke Suzuki.

Invariants of the essential internal methods [1] are enforced for all objects, both ordinary and exotic,
to avoid them having unpredictable behavior that negatively impacts developer experience and creates
security issues.

Enforcing them for Proxy objects is especially cumbersome, given traps are userland code, and quite
expensive performance-wise: invariants invalidation requires checking property attributes,
which currently requires JS => C++ call.

After carefully analyzing all the invariants of the all internal methods, it was concluded that if
[[ProxyTarget]] is an ordinary extensible object without non-configurable properties, validation of
the invariants can be safely (non-observably) skipped.

trap name                   IC    attributes checked
========================    ==    ==================================
getOwnPropertyDescriptor    no    non-configurable || non-extensible
defineProperty              no    non-configurable || non-extensible
has                        yes    non-configurable || non-extensible
get                        yes    non-configurable && (non-writable || accessor)
set                        yes    non-configurable && (non-writable || accessor)
deleteProperty           maybe    non-configurable || non-extensible
ownKeys                     no    non-configurable || non-extensible

This change:

  1. Introduces OverridesIsExtensible out-of-line type info flag to tighten Structure's didPreventExtensions()
     flag, making it more pessimistic: it being `true` doesn't necessarily mean the structure is non-extensible,
     but the reverse of it being `false` can be safely relied on.

     It's not a completely new approach: https://github.com/WebKit/WebKit/pull/11843 does the same thing.

     This change was proved not to introduce bugs since Structure's didPreventExtensions() flag is currently
     checked only for objects without isExtensible() override.

  2. Introduces hasNonConfigurableProperties and hasNonConfigurableReadOnlyOrGetterSetterProperties Structure
     flags that are also pessimistic.

     When inferring their initial values, we do extra care not to be overly pessimistic for objects except
     JSFinalObject by checking overridesGetOwnPropertySlot() so that userland code, which heavily uses
     Map / Set as [[ProxyTarget]], more performant, while making an exception for JSArray's non-configurable
     non-structure "length" property.

     While there is existing hasReadOnlyOrGetterSetterPropertiesExcludingProto flag, we can't combine it with
     hasNonConfigurableProperties because MobX installs a non-writable administration property even on arrays,
     which have non-configurable "length" already, so combining them would disable optimization for JSArray.

  3. Introduces HasStructureWithFlags bytecode op to be utilized in ProxyObject IC helpers to skip trap result
     validation for most of ProxyObject's targets. In the follow-ups, this approach will be expanded to C++ code.

Currently, due to lack of proper profiling of GetInternalField results, constant folding for HasStructureWithFlags
doesn't work in microbenchmarks.

                                   ToT                      patch

proxy-has-miss              159.9825+-2.4190     ^     43.1435+-0.5994        ^ definitely 3.7082x faster
proxy-set-miss-handler      158.7617+-2.3615     ^    139.2158+-1.9197        ^ definitely 1.1404x faster
proxy-set                   125.5967+-1.3678     ^     14.4533+-0.1739        ^ definitely 8.6898x faster
proxy-get                   115.8061+-2.2413     ^     14.2136+-0.1138        ^ definitely 8.1475x faster

<geometric>                 138.5985+-0.9526     ^     33.3237+-0.2255        ^ definitely 4.1592x faster

[1]: https://tc39.es/ecma262/#sec-invariants-of-the-essential-internal-methods

* Source/JavaScriptCore/builtins/BuiltinNames.h:
* Source/JavaScriptCore/builtins/ProxyHelpers.js:
(linkTimeConstant.performProxyObjectHas):
(linkTimeConstant.performProxyObjectGet):
(linkTimeConstant.performProxyObjectSetSloppy):
(linkTimeConstant.performProxyObjectSetStrict):
* Source/JavaScriptCore/bytecode/BytecodeIntrinsicRegistry.h:
* Source/JavaScriptCore/bytecode/BytecodeList.rb:
* Source/JavaScriptCore/bytecode/BytecodeUseDef.cpp:
(JSC::computeUsesForBytecodeIndexImpl):
(JSC::computeDefsForBytecodeIndexImpl):
* Source/JavaScriptCore/bytecode/LinkTimeConstant.h:
* Source/JavaScriptCore/bytecompiler/BytecodeGenerator.cpp:
(JSC::BytecodeGenerator::emitHasStructureWithFlags):
* Source/JavaScriptCore/bytecompiler/BytecodeGenerator.h:
* Source/JavaScriptCore/bytecompiler/NodesCodegen.cpp:
(JSC::BytecodeIntrinsicNode::emit_intrinsic_mustValidateResultOfProxyGetAndSetTraps):
(JSC::BytecodeIntrinsicNode::emit_intrinsic_mustValidateResultOfProxyTrapsExceptGetAndSet):
* Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
(JSC::DFG::AbstractInterpreter<AbstractStateType>::executeEffects):
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::parseBlock):
* Source/JavaScriptCore/dfg/DFGClobberize.h:
(JSC::DFG::clobberize):
* Source/JavaScriptCore/dfg/DFGConstantFoldingPhase.cpp:
(JSC::DFG::ConstantFoldingPhase::foldConstants):
* Source/JavaScriptCore/dfg/DFGDoesGC.cpp:
(JSC::DFG::doesGC):
* Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
(JSC::DFG::FixupPhase::fixupNode):
* Source/JavaScriptCore/dfg/DFGHeapLocation.cpp:
(WTF::printInternal):
* Source/JavaScriptCore/dfg/DFGHeapLocation.h:
* Source/JavaScriptCore/dfg/DFGNode.h:
(JSC::DFG::Node::structureFlags):
* Source/JavaScriptCore/dfg/DFGNodeType.h:
* Source/JavaScriptCore/dfg/DFGPredictionPropagationPhase.cpp:
* Source/JavaScriptCore/dfg/DFGSafeToExecute.h:
(JSC::DFG::safeToExecute):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.h:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT32_64.cpp:
(JSC::DFG::SpeculativeJIT::compile):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
(JSC::DFG::SpeculativeJIT::compile):
* Source/JavaScriptCore/ftl/FTLCapabilities.cpp:
(JSC::FTL::canCompile):
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileNode):
(JSC::FTL::DFG::LowerDFGToB3::compileCompareStrictEq):
* Source/JavaScriptCore/jit/JIT.cpp:
(JSC::JIT::privateCompileMainPass):
* Source/JavaScriptCore/jit/JIT.h:
* Source/JavaScriptCore/jit/JITOpcodes.cpp:
(JSC::JIT::emit_op_has_structure_with_flags):
* Source/JavaScriptCore/llint/LowLevelInterpreter32_64.asm:
* Source/JavaScriptCore/llint/LowLevelInterpreter64.asm:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::init):
* Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::globalFuncHandleProxySetTrapResult): Deleted.
* Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.h:
* Source/JavaScriptCore/runtime/JSTypeInfo.h:
(JSC::TypeInfo::overridesIsExtensible const):
* Source/JavaScriptCore/runtime/PropertySlot.h:
* Source/JavaScriptCore/runtime/ProxyObject.cpp:
(JSC::ProxyObject::performPut):
(JSC::ProxyObject::validatePositiveSetTrapResult):
(JSC::ProxyObject::validateSetTrapResult): Deleted.
* Source/JavaScriptCore/runtime/ProxyObject.h:
* Source/JavaScriptCore/runtime/Structure.cpp:
(JSC::Structure::validateFlags):
(JSC::Structure::Structure):
(JSC::Structure::nonPropertyTransitionSlow):
* Source/JavaScriptCore/runtime/Structure.h:
(JSC::Structure::isNonExtensibleOrHasNonConfigurableProperties const):
(JSC::Structure::bitField const):
* Source/JavaScriptCore/runtime/StructureInlines.h:
(JSC::Structure::add):
(JSC::Structure::attributeChange):
* Source/JavaScriptCore/wasm/js/WebAssemblyGCObjectBase.h:
* Source/WebCore/bindings/js/JSDOMWindowProperties.h:

Canonical link: https://commits.webkit.org/263443@main
